### PR TITLE
update FsLexYacc to 8.0.1 source (by direct copy)

### PR DESCRIPTION
**Pull request by dsyme**
I believe this should fix bug #6346, as discussed in the comments there.  It gets rid of the reliance on .NET taking all mutually-recursive tailcalls for basic lexing.  It should also improve the performance of the lexer a bit.

Note this is a direct-copy update of the files in "buildtools" to match FsLexYacc 8.0.1

### CODE REVIEW

**Comment by dsyme** [comment]
@KevinRansom We should consider how we keep the `buildtools` files in sync with FsLexYacc - we added the source to the nuget package to allow this, but after recent build updates to remove packages.config in `visualfsharp` we don't acquire or record the package version we have copied over at all....

File: src/buildtools/fsyacc/fsyaccast.fs [hunk 1]
```diff
@@ -39,7 +39,7 @@ type Symbols = Symbol list
 //---------------------------------------------------------------------
 // Output Raw Parser Spec AST
 
-let StringOfSym sym = match sym with Terminal s -> "'" + s + "'" | NonTerminal s -> s
+let StringOfSym sym = match sym with Terminal s -> "'" ^ s ^ "'" | NonTerminal s -> s
```
Comment (auduchinok) [inline src/buildtools/fsyacc/fsyaccast.fs]: Is there a reason `^` is preferable?

File: src/buildtools/fslex/fslex.fs [hunk 2]
```diff
@@ -13,7 +13,7 @@ open System.Collections.Generic
 open System.IO 
 
 //------------------------------------------------------------------
-// This code is duplicated from FSharp.Compiler.UnicodeLexing
+// This code is duplicated from Microsoft.FSharp.Compiler.UnicodeLexing
```
Comment (baronfel) [inline src/buildtools/fslex/fslex.fs]: missed a namespace change here

File: src/buildtools/fsyacc/fsyacc.fs [hunk 3]
```diff
@@ -12,7 +12,7 @@ open FsLexYacc.FsYacc
 open FsLexYacc.FsYacc.AST
 
 //------------------------------------------------------------------
-// This code is duplicated from FSharp.Compiler.UnicodeLexing
+// This code is duplicated from Microsoft.FSharp.Compiler.UnicodeLexing
```
Comment (baronfel) [inline src/buildtools/fsyacc/fsyacc.fs]: missed a namespace change here

File: src/buildtools/fsyacc/fsyaccast.fs (hunk 1 above)
Comment (cartermp) [reply src/buildtools/fsyacc/fsyaccast.fs]: I don't think it is, it's just olf F# code

Comment (dsyme) [reply src/buildtools/fsyacc/fsyaccast.fs]: This is copying over the canonical source in FsLexYacc repo - someone must have changed it in visualfsharp repo though they shouldn't have

File: src/buildtools/fslex/fslex.fs (hunk 2 above)
Comment (dsyme) [reply src/buildtools/fslex/fslex.fs]: This is just copying over the canonical source code from FsLexYacc, we'd need to make the change there

File: src/buildtools/fsyacc/fsyacc.fs (hunk 3 above)
Comment (dsyme) [reply src/buildtools/fsyacc/fsyacc.fs]: This is just copying over the canonical source code from FsLexYacc, we'd need to make the change there

**Comment by dsyme** [comment]
This is ready.

**Comment by baronfel** [comment]
As soon as this merges I'll start flowing a new build/release of FCS/FSAC that we can test the fix in editor tooling on nonwindows

File: src/buildtools/fslex/fslex.fs [hunk 4]
```diff
@@ -173,52 +172,51 @@ let main(args: string[]) =
...
-        cfprintfn os "(* Rule %s *)" ident;
-        cfprintfn os "and %s %s (lexbuf : %s.LexBuffer<_>) = _fslex_%s %s %d lexbuf" ident (String.Join(" ",Array.ofList args)) lexlib ident (String.Join(" ",Array.ofList args)) startNode.Id;
-    for ((startNode, actions),(ident,args,_)) in List.zip perRuleData spec.Rules do
-        cfprintfn os "(* Rule %s *)" ident;
-        cfprintfn os "and _fslex_%s %s _fslex_state lexbuf =" ident (String.Join(" ",Array.ofList args));
-        cfprintfn os "  match _fslex_tables.Interpret(_fslex_state,lexbuf) with" ;
-        actions |> Seq.iteri (fun i (code,pos) -> 
-            cfprintfn os "  | %d -> ( " i;
-            cfprintfn os "# %d \"%s\"" pos.Line pos.FileName;
+        cfprintfn os "// Rule %s" ident
```
Comment (TIHan) [inline src/buildtools/fslex/fslex.fs]: This must be the change where it fixes the issue.

Comment (dsyme) [reply src/buildtools/fslex/fslex.fs]: Yes that's right

**Comment by dsyme** [comment]
@TIHan Can we have an approve on this please? Thanks

**Comment by alfonsogarciacaro** [comment]
I tried cherry-picking Microsoft/visualfsharp@e5fcd48 into the FCS fork we're using, did a clean build and copied the FSharp.Compiler.Service.dll to Fable. But I'm still getting SO exceptions when compiling big projects (FCS+Fable itself) on macOS :/

**Comment by dsyme** [comment]
> I tried cherry-picking e5fcd48 into the FCS fork we're using, did a clean build and copied the FSharp.Compiler.Service.dll to Fable. But I'm still getting SO exceptions when compiling big projects (FCS+Fable itself) on macOS :/

Right, but is it the same SO stack?

If fewer tailcalls are being taken then that will always cause problems on bigger projects.  This rooted out one specific problem, but I expect you'll see other ones.  We can _possibly_ make progress one by one, but it would be much better to have the CLR honour the tailcall attributes consistently.

**Comment by alfonsogarciacaro** [comment]
Not sure, don't know how to get the dump on macOS (I don't see the error on Windows). It happens when calling [ParseAndCheckProject](https://github.com/fable-compiler/Fable/blob/0f02fcf42c28764f66f43347dc378f18e33ad16b/src/Fable.Cli/Agent.fs#L96) so I assume this is coming from FCS. On the Fable side, @ncave added a trampoline some time ago to avoid excessive recursion and it's been working so far.

I've been switching the last months between mac and Windows for Fable development so I cannot recall when this started to happen. In CI, Fable only builds itself in Appveyor (Windows) so we haven't seen the issue.

**Pull request merged**